packages/vite/src/node/plugins/clientInjections.ts

```diff
@@ -31,11 +31,12 @@ export function clientInjectionsPlugin(config: ResolvedConfig): Plugin {
         const protocol = hmrConfig?.protocol || null
         const timeout = hmrConfig?.timeout || 30000
         const overlay = hmrConfig?.overlay !== false
+        const isHmrServerSpecified = !!hmrConfig?.server
 
         // hmr.clientPort -> hmr.port
-        // -> (24678 if middleware mode) -> new URL(import.meta.url).port
+        // -> (24678 if middleware mode and HMR server is not specified) -> new URL(import.meta.url).port
         let port = hmrConfig?.clientPort || hmrConfig?.port || null
-        if (config.server.middlewareMode) {
+        if (config.server.middlewareMode && !isHmrServerSpecified) {
```

Comment:
Note: Here we keep port = `null` instead of assigning `24678` when HMR server is specified via config. When the HMR server is specified, HMR WebSocket host (both hostname and port) will be the same as the main server (the one serving HTML and static assets). In this case, we want to rely on ` importMetaUrl.port` [here](https://github.com/vitejs/vite/blob/13ac37df8e135fe6b4990fac6f149a36a19e5409/packages/vite/src/client/client.ts#L29-L29). Before this MR, `24678` was forcefully assigned [here](https://github.com/vitejs/vite/blob/13ac37df8e135fe6b4990fac6f149a36a19e5409/packages/vite/src/client/client.ts#L27-L27)